### PR TITLE
Added (and set as default) a cURL-based method for fetching HTTP content

### PR DESCRIPTION
**Pull request by julienCXX**
Implements proposal in #531.

### CODE REVIEW

**Comment by ArthurHoaro** [comment]
Thanks!

A few remarks:

  * unit tests are failing because PHP changed `CURLOPT_PROGRESSFUNCTION` in v5.5. Not sure what's the best way to handle this. Ping #599.

> 5.5.0 	Added the cURL resource as the first argument to the CURLOPT_PROGRESSFUNCTION callback.

  * the documentation should be updated in Github's wiki, it'll be included in the repo at the next release.
  * it looks like you're downloading the remote page in every case, while the original function only download it if it find a 200 HTTP code. That's not a bad idea, because Shaarli will only make 1 request and users don't usually shaare error pages, however both method should have the same behaviour.

**Comment by julienCXX** [comment]
I think it is fixed now, for the first remark.

For the latter one, I can return an empty body on 40x HTTP codes, even though the body would be still downloaded.
Alternatively, I can make cURL stop immediately when it encounters one of the said codes. But the original headers would not be returned.
Finally, I could send 2 requests, in the fashion of the fallback method, but that would make the code less readable.

**Comment by ArthurHoaro** [comment]
Nah, I like what you did, removing the `or code = 200` [here](https://github.com/julienCXX/Shaarli/blob/39428c7fa93ec1a77d07d29272919059d7a2c784/application/HttpUtils.php#L165) is a better idea, IMHO.

File: application/HttpUtils.php [hunk 1]
```diff
@@ -20,26 +21,136 @@
...
         return array(array(0 => 'Invalid HTTP Url'), false);
     }
 
+    $userAgent = 'Mozilla/5.0 (X11; Fedora; Linux x86_64; rv:45.0)'
+                 . ' Gecko/20100101 Firefox/45.0';
+    $acceptLanguage = substr(setlocale(LC_COLLATE, 0), 0, 2)
+                      . ',en-US;q=0.7,en;q=0.3';
+
+    if (!function_exists('curl_init')) {
+        return get_http_response_fallback($cleanUrl, $timeout, $maxBytes,
```
Comment (ArthurHoaro) [inline application/HttpUtils.php]: If the arguments list is too long, we usually use one line per argument, as stated in [PSR-2](http://www.php-fig.org/psr/psr-2/#4-4-method-arguments).

File: application/HttpUtils.php [hunk 2]
```diff
@@ -20,26 +21,136 @@
...
+        return array(array(0 => 'curl_init() error'), false);
+    }
+
+    // General cURL settings
+    curl_setopt($ch, CURLOPT_AUTOREFERER,       true);
+    curl_setopt($ch, CURLOPT_FOLLOWLOCATION,    true);
+    curl_setopt($ch, CURLOPT_HEADER,            true);
+    curl_setopt($ch, CURLOPT_HTTPHEADER,        array('Accept-Language: '
+                                                      . $acceptLanguage));
+    curl_setopt($ch, CURLOPT_MAXREDIRS,         3); // TODO: set as parameter?
```
Comment (ArthurHoaro) [inline application/HttpUtils.php]: Not necessarily as an argument, but at least as a variable to make sure we use the same value in the fallback function.

File: application/HttpUtils.php [hunk 3]
```diff
@@ -20,26 +21,136 @@
...
+        }
+        return array(array(0 => 'curl_exec() error: ' . $errorStr), false);
+    }
+
+    // Formatting output like the fallback method
+    $rawHeaders = substr($response, 0, $headSize);
+
+    // Keep only headers from latest redirection
+    $rawHeadersArrayRedirs = explode("\r\n\r\n", trim($rawHeaders));
+    $rawHeadersLastRedir = end($rawHeadersArrayRedirs);
```
Comment (ArthurHoaro) [inline application/HttpUtils.php]: Are redirections always the last headers to be set?

Comment (julienCXX) [inline application/HttpUtils.php]: Yes, the latest header set come from the most recent redirect, e.g.:
```
original link headers
<blank line>
1st redirect headers
<blank line>
2nd redirect headers
<blank line>
```

**Comment by julienCXX** [comment]
> Nah, I like what you did, removing the or code = 200 here is a better idea, IMHO.

OK to do that, but I would remove the condition on line 159 too.

**Comment by ArthurHoaro** [comment]
No. We make an extra request changing one parameter if we don't get a HTTP 200. However, if we already have a valid response, the extra request is not necessary.

**Comment by ArthurHoaro** [comment]
> I also noticed that get_headers() follows redirects by default, thus get_redirected_headers() might be useless.

You're right, I don't know how I didn't see that. A bit of clean up could be done later. 

I'll run a few tests but I'm OK with the current PR. Could you squash your commits and update the wiki?

**Comment by julienCXX** [comment]
> Could you squash your commits and update the wiki?

Yes, I can. Which commit structure do you want?

**Comment by ArthurHoaro** [comment]
You changed one file for one feature, I'd say one commit is fine.

**Comment by ArthurHoaro** [comment]
Tested, it works fine.

File: application/HttpUtils.php [hunk 4]
```diff
@@ -20,38 +21,177 @@
...
+
+/**
+ * GET an HTTP URL to retrieve its content (fallback method)
+ *
+ * @param string $cleanUrl       URL to get (http://... valid and in ASCII form)
+ * @param int    $timeout        network timeout (in seconds)
+ * @param int    $maxBytes       maximum downloaded bytes
+ * @param string $userAgent      "User-Agent" header
+ * @param string $acceptLanguage "Accept-Language" header
+ * @param int    $maxRedr        maximum amonut of redirections followed
```
Comment (virtualtam) [inline application/HttpUtils.php]: *amount

**Comment by virtualtam** [comment]
+1 for merging, PR tested with a few known-to-be-problematic web sites listed in #531
(sorry, had to find a small typo somewhere :D )

Btw, this list of problematic websites could be used to add functional (vs. unitary) test suites, to check Shaarli code against real-world conditions. Let me know if there's interest, this would be quite easy to add through PHPUnit's configuration.

**Comment by ArthurHoaro** [comment]
@virtualtam It would be interesting indeed. However, aren't functional tests supposed to test the application a whole, meaning fire up a web server, login, etc., instead of components?

@julienCXX Thanks again!

**Comment by virtualtam** [comment]
Well, you can test both; usually functional tests cover a whole feature in real-world conditions, but can also be used to check features that cannot be covered through unitary tests, e.g. accessing external services. We actually do have such tests, I think most of them are related to `HttpUtils`.

**Pull request merged**